File: ruby/lib/runapi/veo_3_1/resources/extend_video.rb
# frozen_string_literal: true

module RunApi
  module Veo31
    module Resources
      # Veo 3.1 video extension resource.
      # Append additional footage to a completed text-to-video or extend-video task.
      class ExtendVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/veo_3_1/extend_video"

        RESPONSE_CLASS = Types::ExtendVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedExtendVideoResponse

        def initialize(http)
          @http = http
        end

        # Extend a video and wait until complete.
        #
        # @param params [Hash] extend-video parameters (requires source_task_id and prompt)
        # @return [RunApi::Veo31::Types::CompletedExtendVideoResponse] completed task with videos
        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        # Create a video extension task without waiting.
        #
        # @param params [Hash] extend-video parameters
        # @return [RunApi::Veo31::Types::ExtendVideoResponse] task creation result with id
        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        # Get video extension task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Veo31::Types::ExtendVideoResponse] current task status
        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          raise Core::ValidationError, "source_task_id is required" unless params[:source_task_id] || params["source_task_id"]
          raise Core::ValidationError, "prompt is required" unless params[:prompt] || params["prompt"]
        end
      end
    end
  end
end

File: python/src/runapi/veo_3_1/resources/extend_video.py
"""Veo 3.1 video extension resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, ValidationError, RequestOptions

from ..types import CompletedExtendVideoResponse, ExtendVideoResponse


class ExtendVideo(Resource):
    """Extend a previously generated video by referencing its task id."""

    ENDPOINT = "/api/v1/veo_3_1/extend_video"

    RESPONSE_CLASS = ExtendVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedExtendVideoResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Extend a video and poll until it completes.

        Args:
            **params: extend-video parameters (model, ...).

        Returns:
            The completed (narrowed) extend-video response.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create an extend-video task and return immediately with an id.

        Args:
            **params: extend-video parameters (model, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of an extend-video task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        if not params.get("source_task_id"):
            raise ValidationError("source_task_id is required")
        if not params.get("prompt"):
            raise ValidationError("prompt is required")
